File: examples/features/json-transform/python/cott_bindings/curriculum/json_transform/extract_string_field.py
from __future__ import annotations

from cott_runtime import Err, JsonObject, JsonString, JsonValue, Ok, Result
from curriculum.json_transform_types import (
    JsonTransformError,
    JsonTransformError_MissingField,
    JsonTransformError_NotAnObject,
)

def extract_string_field(payload: JsonValue, field: str) -> Result[str, JsonTransformError]:
    if not isinstance(payload, JsonObject):
        return Err(error=JsonTransformError_NotAnObject())
    raw_val = payload.value.get(field)
    if raw_val is None or not isinstance(raw_val, JsonString):
        return Err(error=JsonTransformError_MissingField(field_name=field))
    return Ok(value=raw_val.value)

File: examples/kotlin/features/json-transform/kotlin/cott_impl/curriculum/json_transform/extract_string_field.kt
package cott_impl.curriculum.json_transform

internal fun extract_string_field(payload: cott_runtime.JsonValue, `field`: kotlin.String): cott_runtime.CottResult<curriculum.json_transform.StringField, curriculum.json_transform.JsonTransformError> {
    if (payload !is cott_runtime.JsonObject) {
        return cott_runtime.Err(curriculum.json_transform.JsonTransformError.NotAnObject)
    }
    val member = payload.value[`field`]
    return if (member is cott_runtime.JsonString) {
        cott_runtime.Ok(curriculum.json_transform.StringField(`field`, member.value))
    } else {
        cott_runtime.Err(curriculum.json_transform.JsonTransformError.MissingField(`field`))
    }
}
